text/0002-sarif-result-schema.md

```diff
@@ -0,0 +1,482 @@
+
+    /**
+     * A message that describes the result. The first sentence of the message only will be displayed when visible space
+     * is limited.
+     *
+     * CHECKUP: if we wanted to keep concepts of Result type `SummaryResult | MultiValueResult | LookupValueResult;`, we can store those here, as well as the dataSummary and data for each Result
+     */
+    message: Message;
+
+    /**
+     * A positive integer specifying the number of times this logically unique result was observed in this run.
+     */
+    occurrenceCount?: number;
+
+    /**
+     * A number representing the priority or importance of the result.
+     */
+    rank?: number;
+
+    /**
+     * The stable, unique identifier of the rule, if any, to which this notification is relevant. This member can be
+     * used to retrieve rule metadata from the rules dictionary, if it exists.
+     */
+    ruleId?: string;
+
+    /**
+     * The index within the tool component rules array of the rule object associated with this result.
+     */
+    ruleIndex?: number;
+
+    /**
+     * Key/value pairs that provide additional information about the result.
+     * 
+     */
+    properties?: PropertyBag;
+}
+
+/**
+ * Encapsulates a message intended to be read by the end user.
+ */
+export interface Message {
+    /**
+     * An array of strings to substitute into the message string.
+     */
+    arguments?: string[];
+
+    /**
+     * The identifier for this message.
+     */
+    id?: string;
+
+    /**
+     * A Markdown message string.
+     */
+    markdown?: string;
+
+    /**
+     * A plain text message string.
+     */
+    text?: string;
+
+    /**
+     * Key/value pairs that provide additional information about the message.
+     */
+    properties?: PropertyBag;
+}
+
+/**
+ * A proposed fix for the problem represented by a result object. A fix specifies a set of artifacts to modify. For
+ * each artifact, it specifies a set of bytes to remove, and provides a set of new bytes to replace them.
+ */
+export interface Fix {
+    /**
+     * One or more artifact changes that comprise a fix for a result.
+     */
+    artifactChanges: ArtifactChange[];
+
+    /**
+     * A message that describes the proposed fix, enabling viewers to present the proposed change to an end user.
+     */
+    description?: Message;
+
+    /**
+     * Key/value pairs that provide additional information about the fix.
+     */
+    properties?: PropertyBag;
+}
+
+
+```
+Note: there are type definitions for SARIF [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/289ba098a611b80ee03ed72a2feee41a593fe765/types/sarif/index.d.ts). These types are being used by a number of the high profile tools I linked above, so I view them as very reliable, even though they are third party.
+
+## Open questions: 
+1) Where is it best to store TaskErrors? Options are: properties on `Invocation`, properties on `Run`. I don't see a more specific place in SARIF schema to put nonterminal errors
```

Comment:
I think errors should be stored in the `Run` propertyBag. This will keep it on-par with the existing schema, where the errors property is a sibling of the results.

Reply:
Are `TaskErrors` "meta" errors (failures of the task itself rather than the code)? If so, then `invocation.toolExecutionNotifications` seems like the right place for them to live rather than any custom property bag. See the [SARIF tutorials](https://github.com/microsoft/sarif-tutorials/blob/master/docs/3-Beyond-basics.md#tool-execution-notification-and-tool-configuration-notifications)

Reply:
They are. This is a great suggestion, @drewhannay. 

@carakessler, let's make that change.

Reply:
Good call - thank you @drewhannay!